### ISSUE TITLE
Handle token search parameters in main resource tables

### ISSUE THREAD

**Issue by mattlong** [medplum/medplum#5676]
For query performance and minimizing transactional serialization errors reasons, we intend on moving away from/deprecating the `<Resource>_Token` tables and instead adding columns to the corresponding main `<Resource>` tables to handle searching/filtering by token search parameters.

### Related projects (completed)
To facilitate this project, improvements were made to the `migrate.ts` schema migration generation script. The most of important of which are being able to declaratively define expression-based indexes, declare multiple indexes/columns per `SearchParameter`, and handle derived search parameters. This work has already been shipped mostly in #5686.

Some other closely related changes are to allow more easily defining and using various search parameter implementation strategies and moving code around such that there is a single source of truth for which implementation is being used. This was done in #5672.

### Writing resources and indexing strategy

The proposed new strategy for handling token search parameters consists of adding an array column per search parameter to the `<Resource>` table of type `TEXT[] NOT NULL DEFAULT ARRAY[]::TEXT[]`. Several derived strings per token are added to the array to facilitate the various query patterns possible on tokens.

```ts
interface Token {
  readonly code: string;
  readonly system: string | undefined;
  readonly value: string | undefined;
}

// These code points are reserved and cannot appear in token `system` or `value` values.
const DELIM = '\x01';
const NULL_SYSTEM = '\x02';

// Simplified example code to be added in the `buildColumn()`
// method of `packages/server/src/fhir/repo.ts`
const tokenSet = new Set<string>();
for (const token of buildTokensForSearchParameter(allTokens, resource, searchParam)) {
  const { code, system, value } = token;
  if (system) {
    // [parameter]=[system]|
    tokenSet.add(system);

    if (value) {
      // [parameter]=[system]|[code]
      tokenSet.add(system + DELIM + value);
    }
  }

  if (value) {
    // [parameter]=[code]
    tokenSet.add(DELIM + value);
  
    if (!system) {
      // [parameter]=|[code]
      tokenSet.add(NULL_SYSTEM + DELIM + value);
    }
  }
}

// populated the token's array column
columns[details.columnName] = Array.from(rowTokens);
```

To handle queries efficiently, two indexes are created per token column where `token_array_to_text` is defined as below using another reserved code point as a delimiter:

```sql
CREATE OR REPLACE FUNCTION token_array_to_text(text[])
 RETURNS text
 LANGUAGE sql
 IMMUTABLE
AS $function$SELECT e'\x03'||array_to_string($1, e'\x03')||e'\x03'$function$
```

`"<ResourceType>_<columnName>_idx" gin (<columnName>)`

A GIN index facilitates simple inclusion queries for the various string combos of `system` and `value` shown above in the example code.

`"<ResourceType>_<columnName>Trgm_idx" gin (token_array_to_text(<columnName>) gin_trgm_ops)`

A trigram GIN index facilitates `:text` and `:contains` partial match search operators using `LIKE`, `ILIKE`, and regex Postgres operations.

### Sorting by token values

[R4 spec behavior](https://www.hl7.org/fhir/r4/search.html#_sort):
> A search parameter can refer to an element that repeats, and therefore there can be
multiple values for a given search parameter for a single resource. In this case,
the sort is based on the item in the set of multiple parameters that comes earliest in
the specified sort order when ordering the returned resources.

In [R5](https://www.hl7.org/fhir/r5/search.html#_sort) and beyond, that language is replaced with:
> Servers have discretion on the implementation of sorting for both repeated elements and complex
elements. For example, if requesting a sort on Patient.name, servers might search by family name
then given, given name then family, or prefix, family, and then given. Similarly, when sorting with
multiple given names, the sort might be based on the 'earliest' name in sort order or the first name
in the instance.

With the implementation as described above, sorting would be done over the first value found in the token array column using a Postgres expression like `substring(token_array_to_text(<columnName>), <needle>)` which can result in incorrect result ordering when a resource has multiple token values for the same code. We could consider sorting the order of entries in the array column, split, reverse, and join the array entries back together to correctly handle ascending and descending sorts, but that would be very computationally, very likely prohibitively, expensive at query time.

Alternatively, to correctly handle ascending and descending sorts performantly, two additional `text` columns could be added per search parameter, e.g. `<code>Asc` and `<code>Desc` for each token search parameter that contain the lexically minimum and maximum values respectively. 

### Supporting the `:contains` search modifier

Since trigram indexes are somewhat expensive relative to the amount data being indexed, we should ideally only include tokens that require partial matching in a trigram index. By [the spec](https://hl7.org/fhir/r4/search.html#token), only the `:text` modifier needs to be supported which would cover a subset of token values. However, Medplum [also supports](https://github.com/medplum/medplum/blob/d4b408d7dfa3770453f1afc50c7540a2245bb3c5/packages/server/src/fhir/lookups/token.ts#L554-L556) the `:contains` search modifier on token values even though that is not part of the FHIR spec. To continue supporting `:contains` searches on tokens in a performant way, *all* tokens need to be included in the trigram index.

If we choose NOT to optimize support for `:contains` on arbitrary tokens, we should alter the plan described above by creating an additional column that only contains tokens that would be considered by `:text` searches and ONLY create a trigram index on that column.

**Comment by mattwiller** [comment]
I'd opt to remove the erroneous support for `:contains` on tokens, to prevent index bloat and associated performance losses.  For sorting, would it be okay for us to simply sort in both directions on the first value in the resource itself / the first value in the array column?

**Comment by mattlong** [comment]
@mattwiller Yes, sorting on the first value seems fine to me since there is a lot of discretion on server implementations. The question is really if we want to go above and beyond that so that we aren't returning surprising, unexpected sort orders in results.

**Comment by mattlong** [comment]
Some updates:

## Sorting by token values

We are moving ahead with the simpler implementation where sorting is arbitrarily by the first value encountered found in the array for a given token code.

## Schema design

There is ongoing research at determining what exactly the schema will look like. Instead of a `text[]` column per token search parameter as described above, we are also evaluating a single `text[]` column to hold the derived search strings (with the tokens' code incorporated into the search string) for all of a resource's token search parameters.

Despite the single-column approach needing to contain more text in aggregate (since the token `code` must be repeatedly included in each of the up to to four search strings per token value), a single trigram index on one column with many more array entries uses significantly less disk (almost half in some tests on production-like tables) than individual trigram indexes per token column. This makes sense when you consider the contents of a trigram index Any given trigram, e.g. `abc` is only included once with the list of TIDs it is contained in versus being represented up to N times with each TID also being included up to N times if the trigram were to appear in each token column. The tradeoff being that the single, larger trigram index could potentially be slower than an index per column, but further testing is to be performed to validate that.

## Supporting the `:contains` search modifier for all tokens

Ongoing discussion.
